ingestion/src/metadata/profiler/metrics/system/system.py

```diff
@@ -58,207 +38,81 @@ def recursive_dic():
-
-    Returns:
-        List[Dict]:
-    """
-    logger.debug(f"Fetching system metrics for {dialect}")
-
-    project_id = session.get_bind().url.host
-    dataset_id = table.__table_args__["schema"]  # type: ignore
-
-    metric_results: List[Dict] = []
-
-    jobs = get_value_from_cache(
-        SYSTEM_QUERY_RESULT_CACHE, f"{Dialects.BigQuery}.{project_id}.{dataset_id}.jobs"
-    )
-
-    if not jobs:
-        cursor_jobs = session.execute(
-            text(
-                JOBS.format(
-                    usage_location=conn_config.usageLocation,
-                    dataset_id=dataset_id,
-                    project_id=project_id,
-                    insert=DatabaseDMLOperations.INSERT.value,
-                    update=DatabaseDMLOperations.UPDATE.value,
-                    delete=DatabaseDMLOperations.DELETE.value,
-                    merge=DatabaseDMLOperations.MERGE.value,
-                )
-            )
-        )
-        jobs = [
-            BigQueryQueryResult(
-                query_type=row.statement_type,
-                timestamp=row.start_time,
-                table_name=row.destination_table,
-                dml_statistics=row.dml_statistics,
-            )
-            for row in cursor_jobs
-        ]
-        set_cache(
-            SYSTEM_QUERY_RESULT_CACHE,
-            f"{Dialects.BigQuery}.{project_id}.{dataset_id}.jobs",
-            jobs,
+T = TypeVar("T")
+
+
+class CacheProvider(ABC, Generic[T]):
+    def __init__(self):
+        self.cache = LRUCache[T](LRU_CACHE_SIZE)
+
+    def get_or_update_cache(
+        self,
+        cache_path: str,
+        get_queries_fn: Callable[..., List[T]],
+        *args,
+        **kwargs,
+    ):
+        if cache_path in self.cache:
+            return self.cache.get(cache_path)
+        result = get_queries_fn(*args, **kwargs)
+        self.cache.put(cache_path, result)
+        return result
+
+
+class EmptySystemMetricsSource:
+    """Empty system metrics source that can be used as a default. Just returns an empty list of system metrics
+    for any resource."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs.pop("session", None)
+        if len(args) > 0:
+            logger.warning("Received unexpected arguments: %s", args)
+        if len(kwargs) > 0:
+            logger.warning("Received unexpected keyword arguments: %s", kwargs)
+        super().__init__()
+
+    def get_inserts(self, *args, **kwargs) -> List[SystemProfile]:
+        """Get insert queries"""
+        return []
+
+    def get_deletes(self, *args, **kwargs) -> List[SystemProfile]:
+        """Get delete queries"""
+        return []
+
+    def get_updates(self, *args, **kwargs) -> List[SystemProfile]:
+        """Get update queries"""
+        return []
+
+    def get_kwargs(self, *args, **kwargs):
+        return {}
+
+
+class SystemMetricsComputer(EmptySystemMetricsSource):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```

Comment:
That will not be needed I believe

Reply:
It's required for using the `super()` dependency injection method. The one downside of this method. I'll add a comment for it.

Reply:
P.S: Something similar is used in [python's ABC class](https://github.com/python/cpython/blob/de0d5c6e2e12f24ade1ccc457afaf5fb2c650c64/Lib/abc.py#L106)

Reply:
That is a bit different. It overrides `__new__` not `__init__`. The idea, and what they do in ABC, is override `__new__` if you need to do something after class instantiation. When you override `__new__` you do need to call `super().__new__` so you instantiate the class -- otherwise you'll get `None` instance. 

```
 
class Foo:
    def __init__(self) -> None:
        self.a = 1
        self.b # 2

    def __new__(cls):
        cls = super().__new__(cls)
        cls.b = 2
        return cls

f = Foo()
f.a # 1
f.b # 2
```

Here we call `super().__init__(...)` but if we don't have that method here it will call the one of the parent.